### DISCUSSION TITLE
raise_for_status should print the content too

### DISCUSSION THREAD

**Discussion by nkitsaini** [encode/httpx#1752] [Ideas]
`raise_for_status` only gives information about status code but many times the response content contains useful information.   
It would be helpful it could also include the `content` in `Exception`.  
 It could be configurable but I can't think of case where I don't want content to be printed in case of failure.

**Comment by lovelydinosaur** [comment]
Well, that depends. The content could be excessively large, so we might(?) need to cap how much we'd include in the exception. Also it could be binary content, and it could be compressed.

A good *starting point* would be to compare if `requests` currently includes the request content in their `raise_for_status` exceptions. If not, then I'm okay with us also not doing so *for the time being*, although it'd certainly be a neat improvement at some point.

**Reply by nkitsaini** [reply]
So `requests` doesn't print the content. So I think it's fine for time being but I would suggest `.ok()` method, so it's easier to do following:
```py
resp = httpx.get(...)
if not resp.ok():
    log.warn("failed", content=resp.text)
```
`.ok()` is also part of requests.
I can create new issue or pull request for this.

**Reply by senpos** [reply]
IIRC, the `Response.ok` property was initially added to `httpx` as a part of compatibility with `requests`, though it was replaced with a bit more clear `is_error` property.

```python
import httpx

response = httpx.get('https://httpbin.org/status/500')
if response.is_error:
	print(f'Error {response.status_code}: {response.text or "<empty>"}')
```

But I agree, it would be nice to have the response body in the exception raised by `raise_for_status` in the future. :-)

**Comment by tyilo** [comment]
Instead of always having it, providing a keyword argument to enable would also be fine with me.
Something like `r.raise_for_status(show_body=True)`

**Comment by bentheiii** [comment]
Any update on this? It would be a massive QoL improvement for users. As of right now my team is forced to consider `raise_for_status` as an antipattern because if something goes wrong, we can't tell what from the error it raises.

**Reply by bentheiii** [reply]
I'd be happy to submit a PR with a maintainer's go-ahead

**Comment by peter-gerlagh-kyos** [comment]
I'm in the same position. `raise_for_status` is convenient, but not being able to see `response.json()` or `response.request.content` in the error log is frustrating. Has there been any progress since 2022?

**Reply by lovelydinosaur** [reply]
On review this is a no from me.

**Comment by thomasborgen** [comment]
Wouldn't a nice middle ground be to allow users to opt in to having the content included with a parameter to `raise_for_status`. something like: `include_content` which has a default value of `False`.

Since the docs is encouraging users to use raise_for_status chained as in the example below, I think that not being able to control this just renders this pattern quite useless.

```
The method returns the response instance, allowing you to use it inline. For example:

>>> r = httpx.get('...').raise_for_status()
>>> data = httpx.get('...').raise_for_status().json()
```
